src/engine/GroupBy.cpp

```diff
@@ -677,6 +697,123 @@ bool GroupBy::computeOptimizedGroupByIfPossible(IdTable* result) {
   }
 }
 
+// _____________________________________________________________________________
+std::optional<GroupBy::ExplicitlySortedData>
+    GroupBy::checkIfExplicitlySorted() {
+  auto* sort = dynamic_cast<const Sort*>(_subtree->getRootOperation().get());
+  if (!sort) {
+    return std::nullopt;
+  }
+
+  if (_groupByVariables.size() != 1) {
+    return std::nullopt;
+  }
+
+  const Variable& groupByVariable = _groupByVariables.front();
+  auto child = _subtree->getRootOperation()->getChildren().at(0);
+  auto columnIndex = child->getVariableColumn(groupByVariable);
+
+  return ExplicitlySortedData{ columnIndex };
+}
+
+// _____________________________________________________________________________
+template <size_t OUT_WIDTH>
+bool GroupBy::computeGroupByForSortedSubtree(IdTable* result,
+                                             std::vector<Aggregate> aggregates,
+                                             const IdTable& subresult,
+                                             size_t columnIndex,
+                                             LocalVocab& localVocab) {
+  // TODO: Use an interface for AggregationData
+  ad_utility::HashMap<ValueId, AverageAggregationData> map;
+
+  // Initialize evaluation context
+  sparqlExpression::EvaluationContext evaluationContext(
+      *getExecutionContext(), _subtree->getVariableColumns(),
+      subresult,getExecutionContext()->getAllocator(), localVocab);
+
+  evaluationContext._groupedVariables = ad_utility::HashSet<Variable>{
+      _groupByVariables.begin(), _groupByVariables.end()};
+  // TODO: Save previous results (?)
+  evaluationContext._variableToColumnMapPreviousResults =
+      getInternallyVisibleVariableColumns();
+  evaluationContext._previousResultsFromSameGroup.resize(getResultWidth());
+  evaluationContext._isPartOfGroupBy = true;
+
+  for (size_t i = 0; i < subresult.size(); ++i) {
+    auto id = subresult(i, columnIndex);
+
+    // We always calculate for a single row
+    evaluationContext._beginIndex = i;
+    evaluationContext._endIndex = i + 1;
```

Comment:
This is probably the reason why it is slow when your child expressions are more complex.
You should evaluate the child expression for a whole block of children at once (make the blocksize a variable and play with it,
And then you get basically `n` pairs of `(groupByVariable, subresult)` with a single call to `evaluate()`. For each of those you then do your normal lookup in the hash table.